simde/arm/neon/qshl.h

```diff
@@ -0,0 +1,812 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright:
+ *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2020      Sean Maher <seanptmaher@gmail.com>
+ */
+
+#if !defined(SIMDE_ARM_NEON_QSHL_H)
+#define SIMDE_ARM_NEON_QSHL_H
+
+#include "types.h"
+
+HEDLEY_DIAGNOSTIC_PUSH
+SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+SIMDE_BEGIN_DECLS_
+
+#define SIMDE_MIN_(A, B) (((A) > (B)) ? (B) : (A))
+
+SIMDE_FUNCTION_ATTRIBUTES
+simde_int8x8_t
+simde_vqshl_s8(simde_int8x8_t a, simde_int8x8_t b) {
+  #if defined(SIMDE_ARM_NEON_A32V7_NATIVE)
+    return vqshl_s8(a, b);
+  #else
+    simde_int8x8_private
+      r_,
+      a_ = simde_int8x8_to_private(a),
+      b_ = simde_int8x8_to_private(b);
+
+    SIMDE_VECTORIZE
+    for (size_t i = 0 ; i < (sizeof(r_.values) / sizeof(r_.values[0])) ; i++) {
+      int8_t b_value = HEDLEY_STATIC_CAST(int8_t, b_.values[i] & 0xff); // HEDLEY_STATIC_CAST(int8_t, b_.values[i]);
+      uint8_t mask = UINT8_MAX << SIMDE_MIN_((7 - b_value), 8);
+      if (b_value > 0) {
+        if (b_value >= 8) {
+          if (a_.values[i]) {
```

Comment:
The overall logic is a bit confusing for me. E.g. I don't get why we need to check `if (a_.values[i])`.
If `b_value is >= 8` then we can immediate set `r_.values[i]` to be 0, can't we?

Not trying to add more work for you, maybe you can consider implementing sshl first, then you qshl can be sshl + saturate, which may simpilfy the logic here a bit.
See https://git.linaro.org/arm/vixl.git/tree/src/aarch64/simulator-aarch64.cc#n5301

Reply:
If `b_value is >= 8`, then either we're saturating, or the value was 0 to begin with, so I've gotta make another check.